File: VoiceRecorder/SaveViewModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using VoiceRecorder.Core;
using NAudio.Wave;
using System.Windows.Input;
using Microsoft.Win32;
using System.Windows;
using System.IO;
using VoiceRecorder.Audio;
using VoiceRecorder.Properties;
using GalaSoft.MvvmLight.Command;
using GalaSoft.MvvmLight;
using GalaSoft.MvvmLight.Messaging;

namespace VoiceRecorder
{
    class SaveViewModel : ViewModelBase, IView
    {
        private VoiceRecorderState voiceRecorderState;
        private SampleAggregator sampleAggregator;
        private int leftPosition;
        private int rightPosition;
        private int totalWaveFormSamples;
        private IAudioPlayer audioPlayer;
        private int samplesPerSecond;
        public const string ViewName = "SaveView";

        public SaveViewModel(IAudioPlayer audioPlayer)
        {
            this.SampleAggregator = new SampleAggregator();
            SampleAggregator.NotificationCount = 800; // gets set correctly later on
            this.audioPlayer = audioPlayer;
            this.SaveCommand = new RelayCommand(() => Save());
            this.SelectAllCommand = new RelayCommand(() => SelectAll());
            this.PlayCommand = new RelayCommand(() => Play());
            this.StopCommand = new RelayCommand(() => Stop());
            this.AutoTuneCommand = new RelayCommand(() => OnAutoTune());
            Messenger.Default.Register<ShuttingDownMessage>(this, (message) => OnShuttingDown(message));
        }

        private void OnAutoTune()
        {
            Messenger.Default.Send(new NavigateMessage("AutoTuneView", this.voiceRecorderState));
        }

        public ICommand SaveCommand { get; private set; }
        public ICommand SelectAllCommand { get; private set;  }
        public ICommand PlayCommand { get; private set; }
        public ICommand StopCommand { get; private set; }
        public ICommand AutoTuneCommand { get; private set; }

        public void Activated(object state)
        {
            this.voiceRecorderState = (VoiceRecorderState)state;
            RenderFile();
        }

        private void OnShuttingDown(ShuttingDownMessage message)
        {
            audioPlayer.Dispose();
            
            if (message.CurrentViewName == SaveViewModel.ViewName)
            {
                this.voiceRecorderState.DeleteFiles();
            }
        }

        private void Save()
        {
            SaveFileDialog saveFileDialog = new SaveFileDialog();
            saveFileDialog.Filter = "WAV file (.wav)|*.wav|MP3 file (.mp3)|.mp3";
            saveFileDialog.DefaultExt = ".wav";
            bool? result = saveFileDialog.ShowDialog();
            if (result.HasValue && result.Value)
            {
                SaveAs(saveFileDialog.FileName);
            }
        }

        private TimeSpan PositionToTimeSpan(int position)
        {
            int samples = SampleAggregator.NotificationCount * position;
            return TimeSpan.FromSeconds((double)samples / samplesPerSecond);
        }

        private void SaveAs(string fileName)
        {
            AudioSaver saver = new AudioSaver(voiceRecorderState.ActiveFile);
            saver.TrimFromStart = PositionToTimeSpan(LeftPosition);
            saver.TrimFromEnd = PositionToTimeSpan(TotalWaveFormSamples - RightPosition);

            if (fileName.ToLower().EndsWith(".wav"))
            {
                saver.SaveFileFormat = SaveFileFormat.Wav;
                saver.SaveAudio(fileName);
            }
            else if (fileName.ToLower().EndsWith(".mp3"))
            {
                string lameExePath = LocateLame();
                if (lameExePath != null)
                {
                    saver.SaveFileFormat = SaveFileFormat.Mp3;
                    saver.LameExePath = lameExePath;
                    saver.SaveAudio(fileName);
                }
            }
            else
            {
                MessageBox.Show("Please select a supported output format");
            }
        }

        public int LeftPosition
        {
            get
            {
                return leftPosition;
            }
            set
            {
                if (leftPosition != value)
                {
                    leftPosition = value;
                    RaisePropertyChanged("LeftPosition");
                }
            }
        }

        public int RightPosition
        {
            get
            {
                return rightPosition;
            }
            set
            {
                if (rightPosition != value)
                {
                    rightPosition = value;
                    RaisePropertyChanged("RightPosition");
                }
            }
        }

        public string LocateLame()
        {
            string lameExePath = Settings.Default.LameExePath;

            if (String.IsNullOrEmpty(lameExePath) || !File.Exists(lameExePath))
            {
                if (MessageBox.Show("To save as MP3 requires LAME.exe, please locate",
                    "Save as MP3",
                    MessageBoxButton.OKCancel) == MessageBoxResult.OK)
                {
                    OpenFileDialog ofd = new OpenFileDialog();
                    ofd.FileName = "lame.exe";
                    bool? result = ofd.ShowDialog();
                    if (result != null && result.HasValue)
                    {
                        if (File.Exists(ofd.FileName) && ofd.FileName.ToLower().EndsWith("lame.exe"))
                        {
                            Settings.Default.LameExePath = ofd.FileName;
                            Settings.Default.Save();
                            return ofd.FileName;
                        }
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }
            return lameExePath;
        }

        private void RenderFile()
        {
            SampleAggregator.RaiseRestart();
            using (WaveFileReader reader = new WaveFileReader(this.voiceRecorderState.ActiveFile))
            {
                this.samplesPerSecond = reader.WaveFormat.SampleRate;
                SampleAggregator.NotificationCount = reader.WaveFormat.SampleRate/10;
            
                byte[] buffer = new byte[1024];
                WaveBuffer waveBuffer = new WaveBuffer(buffer);
                waveBuffer.ByteBufferCount = buffer.Length;
                int bytesRead;
                do
                {
                    bytesRead = reader.Read(waveBuffer, 0, buffer.Length);
                    int samples = bytesRead / 2;
                    for (int sample = 0; sample < samples; sample++)
                    {
                        if (bytesRead > 0)
                        {
                            sampleAggregator.Add(waveBuffer.ShortBuffer[sample] / 32768f);
                        }
                    }
                } while (bytesRead > 0);
                int totalSamples = (int)reader.Length / 2;
                TotalWaveFormSamples = totalSamples / sampleAggregator.NotificationCount;
                SelectAll();
            }
            audioPlayer.LoadFile(this.voiceRecorderState.ActiveFile);
        }

        private void Play()
        {
            audioPlayer.StartPosition = PositionToTimeSpan(LeftPosition);
            audioPlayer.EndPosition = PositionToTimeSpan(RightPosition);
            audioPlayer.Play();
        }

        private void Stop()
        {
            audioPlayer.Stop();
        }

        private void SelectAll()
        {
            LeftPosition = 0;
            RightPosition = TotalWaveFormSamples;
        }

        public SampleAggregator SampleAggregator
        {
            get 
            {
                return sampleAggregator;  
            }
            set
            {
                if (sampleAggregator != value)
                {
                    sampleAggregator = value; 
                    RaisePropertyChanged("SampleAggregator");
                }
            }
        }        

        public int TotalWaveFormSamples
        {
            get
            {
                return totalWaveFormSamples;
            }
            set
            {
                if (totalWaveFormSamples != value)
                {
                    totalWaveFormSamples = value;
                    RaisePropertyChanged("TotalWaveFormSamples");
                }
            }
        }
    }
}

File: ~VB Port/VoiceRecorder/SaveViewModel.vb
Imports System.Text
Imports VoiceRecorder.Core
Imports NAudio.Wave
Imports Microsoft.Win32
Imports System.IO
Imports VoiceRecorder.Audio
Imports GalaSoft.MvvmLight.Command
Imports GalaSoft.MvvmLight
Imports GalaSoft.MvvmLight.Messaging
Imports My

Namespace VoiceRecorder
    Friend Class SaveViewModel
        Inherits ViewModelBase
        Implements IView
        Private voiceRecorderState As VoiceRecorderState
        Private _sampleAggregator As SampleAggregator
        Private _leftPosition As Integer
        Private _rightPosition As Integer
        Private _totalWaveFormSamples As Integer
        Private audioPlayer As IAudioPlayer
        Private samplesPerSecond As Integer
        Public Const ViewName = "SaveView"

        Public Sub New(ByVal audioPlayer As IAudioPlayer)
            Me.SampleAggregator = New SampleAggregator
            SampleAggregator.NotificationCount = 800 ' gets set correctly later on
            Me.audioPlayer = audioPlayer
            Me.SaveCommand = New RelayCommand(Sub() Save())
            Me.SelectAllCommand = New RelayCommand(Sub() SelectAll())
            Me.PlayCommand = New RelayCommand(Sub() Play())
            Me.StopCommand = New RelayCommand(Sub() Me.Stop())
            Me.AutoTuneCommand = New RelayCommand(Sub() OnAutoTune())
            Messenger.Default.Register(Of ShuttingDownMessage)(Me, Sub(message) OnShuttingDown(message))
        End Sub

        Private Sub OnAutoTune()
            audioPlayer.Dispose() ' needed to relinquish the file as it may get deleted
            Messenger.Default.Send(New NavigateMessage("AutoTuneView", Me.voiceRecorderState))
        End Sub

        Private _saveCommand As ICommand
        Public Property SaveCommand As ICommand
            Get
                Return _saveCommand
            End Get
            Private Set(ByVal value As ICommand)
                _saveCommand = value
            End Set
        End Property

        Private _selectAllCommand As ICommand
        Public Property SelectAllCommand As ICommand
            Get
                Return _selectAllCommand
            End Get
            Private Set(ByVal value As ICommand)
                _selectAllCommand = value
            End Set
        End Property

        Private _playCommand As ICommand
        Public Property PlayCommand As ICommand
            Get
                Return _playCommand
            End Get
            Private Set(ByVal value As ICommand)
                _playCommand = value
            End Set
        End Property

        Private _stopCommand As ICommand
        Public Property StopCommand As ICommand
            Get
                Return _stopCommand
            End Get
            Private Set(ByVal value As ICommand)
                _stopCommand = value
            End Set
        End Property

        Private _autoTuneCommand As ICommand
        Public Property AutoTuneCommand As ICommand
            Get
                Return _autoTuneCommand
            End Get
            Private Set(ByVal value As ICommand)
                _autoTuneCommand = value
            End Set
        End Property

        Public Sub Activated(ByVal state As Object) Implements IView.Activated
            Me.voiceRecorderState = CType(state, VoiceRecorderState)
            RenderFile()
        End Sub

        Private Sub OnShuttingDown(ByVal message As ShuttingDownMessage)
            audioPlayer.Dispose()

            If message.CurrentViewName = SaveViewModel.ViewName Then
                Me.voiceRecorderState.DeleteFiles()
            End If
        End Sub

        Private Sub Save()
            Dim saveFileDialog As New SaveFileDialog
            saveFileDialog.Filter = "WAV file (.wav)|*.wav|MP3 file (.mp3)|.mp3"
            saveFileDialog.DefaultExt = ".wav"
            Dim result? As Boolean = saveFileDialog.ShowDialog()
            If result.HasValue AndAlso result.Value Then
                SaveAs(saveFileDialog.FileName)
            End If
        End Sub

        Private Function PositionToTimeSpan(ByVal position As Integer) As TimeSpan
            Dim samples = SampleAggregator.NotificationCount * position
            Return TimeSpan.FromSeconds(CDbl(samples) / samplesPerSecond)
        End Function

        Private Sub SaveAs(ByVal fileName As String)
            Dim saver As New AudioSaver(voiceRecorderState.ActiveFile)
            saver.TrimFromStart = PositionToTimeSpan(LeftPosition)
            saver.TrimFromEnd = PositionToTimeSpan(TotalWaveFormSamples - RightPosition)

            If fileName.ToLower().EndsWith(".wav") Then
                saver.SaveFileFormat = SaveFileFormat.Wav
                saver.SaveAudio(fileName)
            ElseIf fileName.ToLower().EndsWith(".mp3") Then
                Dim lameExePath = LocateLame()
                If lameExePath IsNot Nothing Then
                    saver.SaveFileFormat = SaveFileFormat.Mp3
                    saver.LameExePath = lameExePath
                    saver.SaveAudio(fileName)
                End If
            Else
                MessageBox.Show("Please select a supported output format")
            End If
        End Sub

        Public Property LeftPosition As Integer
            Get
                Return _leftPosition
            End Get
            Set(ByVal value As Integer)
                If _leftPosition <> value Then
                    _leftPosition = value
                    RaisePropertyChanged("LeftPosition")
                End If
            End Set
        End Property

        Public Property RightPosition As Integer
            Get
                Return _rightPosition
            End Get
            Set(ByVal value As Integer)
                If _rightPosition <> value Then
                    _rightPosition = value
                    RaisePropertyChanged("RightPosition")
                End If
            End Set
        End Property

        Public Function LocateLame() As String
            Dim lameExePath = Settings.Default.LameExePath

            If String.IsNullOrEmpty(lameExePath) OrElse (Not File.Exists(lameExePath)) Then
                If MessageBox.Show("To save as MP3 requires LAME.exe, please locate",
                                   "Save as MP3", MessageBoxButton.OKCancel) = MessageBoxResult.OK Then
                    Dim ofd As New OpenFileDialog
                    ofd.FileName = "lame.exe"
                    Dim result? = ofd.ShowDialog()
                    If result IsNot Nothing AndAlso result.HasValue Then
                        If File.Exists(ofd.FileName) AndAlso ofd.FileName.ToLower().EndsWith("lame.exe") Then
                            Settings.Default.LameExePath = ofd.FileName
                            Settings.Default.Save()
                            Return ofd.FileName
                        End If
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End If
            Return lameExePath
        End Function

        Private Sub RenderFile()
            SampleAggregator.RaiseRestart()
            Using reader As New WaveFileReader(Me.voiceRecorderState.ActiveFile)
                Me.samplesPerSecond = reader.WaveFormat.SampleRate
                SampleAggregator.NotificationCount = reader.WaveFormat.SampleRate \ 10

                Dim buffer(1023) As Byte
                Dim waveBuffer As New WaveBuffer(buffer)
                waveBuffer.ByteBufferCount = buffer.Length
                Dim bytesRead As Integer
                Do
                    bytesRead = reader.Read(waveBuffer, 0, buffer.Length)
                    Dim samples = bytesRead \ 2
                    For sample = 0 To samples - 1
                        If bytesRead > 0 Then
                            _sampleAggregator.Add(waveBuffer.ShortBuffer(sample) / 32768.0F)
                        End If
                    Next sample
                Loop While bytesRead > 0
                Dim totalSamples = CInt(Fix(reader.Length)) \ 2
                TotalWaveFormSamples = totalSamples \ _sampleAggregator.NotificationCount
                SelectAll()
            End Using
            audioPlayer.LoadFile(Me.voiceRecorderState.ActiveFile)
        End Sub

        Private Sub Play()
            audioPlayer.StartPosition = PositionToTimeSpan(LeftPosition)
            audioPlayer.EndPosition = PositionToTimeSpan(RightPosition)
            audioPlayer.Play()
        End Sub

        Private Sub [Stop]()
            audioPlayer.Stop()
        End Sub

        Private Sub SelectAll()
            LeftPosition = 0
            RightPosition = TotalWaveFormSamples
        End Sub

        Public Property SampleAggregator As SampleAggregator
            Get
                Return _sampleAggregator
            End Get
            Set(ByVal value As SampleAggregator)
                If _sampleAggregator IsNot value Then
                    _sampleAggregator = value
                    RaisePropertyChanged("SampleAggregator")
                End If
            End Set
        End Property

        Public Property TotalWaveFormSamples As Integer
            Get
                Return _totalWaveFormSamples
            End Get
            Set(ByVal value As Integer)
                If _totalWaveFormSamples <> value Then
                    _totalWaveFormSamples = value
                    RaisePropertyChanged("TotalWaveFormSamples")
                End If
            End Set
        End Property
    End Class
End Namespace
